src/query/owner/commands/index.ts

```diff
@@ -0,0 +1,6 @@
+export * from './owner-created.command';
```

Comment:
Huh? Aan de query kant zijn commands toch niet meer aanwezig? Hier zouden alleen events ge'listen'd moeten worden ... 🤔

Reply:
@marcvanandel Ik had expres commando's gebruikt om het geheel meer modulair te maken. Op deze manier hoef je niet een eventhandler te hebben waar alle schrijf logica voor alle events in staat. Ik vond het zelf wel mooi om gewoon een commando af te vuren, en in de commandoHandler alle logica te stoppen zoals de projectie bijwerken / websocket signaal sturen. Ik kan het wel iets anders opzetten zodat niet meer de NestJs commandohandler gebruikt wordt, maar er direct een klasse wordt aangesproken, dat is misschien duidelijker.

Reply:
Ik snap het ook nog niet helemaal. Interpreteer je hier de query folder als een API endpoint waar je deze commands heen kan sturen? Of zijn de commands als reactie op de events die terugkomen uit de EventStore. In dat geval kunnen we wellicht Saga's gebruiken 🙂

Reply:
Als je het [plaatje](https://github.com/kadaster-labs/sensrnet-registry-backend/tree/initial-project#architecture) bekijkt wat ik heb gemaakt, dan zou ik een saga verwachten tussen de http client en de command en query kant. De client communiceert met de saga om commands uit te voeren en om feedback te krijgen als de query kant is bijgewerkt ... Het zou ook kunnen zijn dat een saga gestart wordt vanuit de command kant; dat weet ik niet precies. Sagas zijn best complex en worden verschillend gebruikt in de verschillende frameworks. Ik heb hier vast niet een goed beeld van hoe dat in NestJS dat gebruikelijk is 😁 

De flow is (zou moeten zijn - volgens mij):
http client -> command (api / controller) -> command handler -> aggregate -> events -> eventstore -> event-producer (listener + db-updater) -> database -> query-api (controller + db-lookup)

Reply:
klinkt goed, zo ziet het er nu ook uit, behalve dat er binnen de event-producer gebruikt gemaakt van commando's / commando handlers omdat ik het makkelijk vond om op die manier de logica te scheiden. Ik ga nog even kijken hoe dat beter kan.

Reply:
Klinkt goed, dan zijn saga's hier inderdaad niet nodig. Die worden getriggered door een event en "Using the RxJS library, it can combine, merge, filter or apply other RxJS operators on the event stream" (NestJS CQRS docs) om vervolgens een nieuwe Command af te vuren, wat me voor ons doel niet nodig lijkt.

Reply:
Nee ... behalve als we reactive gaan. Dat is wel een nieuw 'command' richting alle clients dat de view / query kant is bijgewerkt, niet?

Reply:
Ik was net bezig om te kijken hoe we dat kunnen doen. Een optie is dat we subscriben op updates uit de mongo db, met [Change streams](https://mongoosejs.com/docs/models.html#change-streams). Maar daarvoor moet de mongodb achter een replicaset zitten. 

Als saga's dit ook kunnen, gaat het wel mooier zijn om die te gebruiken. Ik ga uitzoeken of we dit daarmee kunnen oplossen 🙂

Reply:
Check. Ik zou verwachten dat een saga getriggered zou kunnen worden vanuit de event-bus of event-processor of zo? Change streams is ook wel mooi, maar een lock-in op Mongo ... welke ik het liefst wil minimaal te houden... Dus EENS! 😀

Reply:
Huh, is niet nodig toch? Ik werk de mongodb bij: https://github.com/kadaster-labs/sensrnet-registry-backend/blob/950d09b4329ef3a126add1db1fe4f53978eb1b7e/src/query/owner/commands/owner-created.handler.ts#L20. Als callback kan je dan een message versturen, wat opgepikt wordt door de NestJs websockets die subscribed zijn aan deze message d.m.v. subscribemessage) (https://docs.nestjs.com/websockets/gateways). De message kan verzonden d.m.v. socket.emit() binnen de mongodb created callback, zie de nestjs documentatie.

Reply:
Dus: mongo wordt geupdaed -> verstuur socket.emit('updated') -> message handlers die subscribed zijn d.m.v. @SubscribeMessage('updated') pushen de wijzigingen naar de clients.

Reply:
Risico dat je dan loopt (ipv een saga), is dat de listener verantwoordelijkheid heeft voor het luisteren naar events + bijwerken van de database (beide OK) én triggeren van messages richting clients. Volgens mij is het beter om dat te ontkoppelen. Stel dat we voorwaarden hebben / krijgen of complexiteit krijgen in het triggeren van clients, dan wil je niet de event-producer (ik blijf 'm nog maar zo noemen) complexer maken. En de database bijwerken zou prima moeten kunnen _zonder_ dat er clients zijn. Dus ook daarom lijken mij dat verschillende verantwoordelijkheden ...

Reply:
Ik heb een voorbeeld bijgevoegd om het iets duidelijker te maken: na de save methode trigger je de WebSocketGateway https://github.com/kadaster-labs/sensrnet-registry-backend/blob/54a43fc008de20e4540f2dc9929c972a19f0c8fa/src/query/owner/commands/owner-created.handler.ts#L25, die dan validatie uitvoert en een notificatie pushed naar de clients: https://github.com/kadaster-labs/sensrnet-registry-backend/blob/54a43fc008de20e4540f2dc9929c972a19f0c8fa/src/query/owner/owner.gateway.ts#L1 Dat is toch wat we zoeken?

Reply:
Ik weet het niet. Voor nu lijkt me dit prima. Laten we maar zorgen dat het gaat werken ... Ik kan me voorstellen dat als we hier meer ervaring mee krijgen en beter begrijpen hoe dit werkt, dat we hier een saga zouden triggeren ipv de sockets gateway ...